File: app/src/test/kotlin/fr/free/nrw/commons/login/UserExtendedInfoClientTest.java
package fr.free.nrw.commons.login;

import android.net.Uri;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.stream.MalformedJsonException;
import fr.free.nrw.commons.MockWebServerTest;
import fr.free.nrw.commons.auth.login.LoginInterface;
import io.reactivex.observers.TestObserver;
import org.junit.Before;
import org.junit.Test;
import org.wikipedia.dataclient.WikiSite;
import org.wikipedia.dataclient.mwapi.MwQueryResponse;
import org.wikipedia.json.NamespaceTypeAdapter;
import org.wikipedia.json.PostProcessingTypeAdapter;
import org.wikipedia.json.UriTypeAdapter;
import org.wikipedia.json.WikiSiteTypeAdapter;
import org.wikipedia.page.Namespace;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

public class UserExtendedInfoClientTest extends MockWebServerTest {

    private LoginInterface apiService;

    @Override
    @Before
    public void setUp() throws Throwable {
        super.setUp();

        apiService = new Retrofit.Builder()
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .addConverterFactory(GsonConverterFactory.create(getGson()))
            .baseUrl(server().getUrl())
            .build()
            .create(LoginInterface.class);
    }

    @Test
    public void testRequestSuccess() throws Throwable {
        enqueueFromFile("user_extended_info.json");
        TestObserver<MwQueryResponse> observer = new TestObserver<>();

        apiService.getUserInfo("USER").subscribe(observer);

        observer
            .assertComplete()
            .assertNoErrors()
            .assertValue(
                result -> result.query().userInfo().id() == 24531888
                && result.query().getUserResponse("USER").name().equals("USER")
            );
    }

    @Test
    public void testRequestResponse404() {
        enqueue404();
        TestObserver<MwQueryResponse> observer = new TestObserver<>();

        apiService.getUserInfo("USER").subscribe(observer);

        observer.assertError(Exception.class);
    }

    @Test
    public void testRequestResponseMalformed() {
        enqueueMalformed();
        TestObserver<MwQueryResponse> observer = new TestObserver<>();

        apiService.getUserInfo("USER").subscribe(observer);

        observer.assertError(MalformedJsonException.class);
    }

    private Gson getGson() {
        return new GsonBuilder()
            .registerTypeHierarchyAdapter(Uri.class, new UriTypeAdapter().nullSafe())
            .registerTypeHierarchyAdapter(Namespace.class, new NamespaceTypeAdapter().nullSafe())
            .registerTypeAdapter(WikiSite.class, new WikiSiteTypeAdapter().nullSafe())
            .registerTypeAdapterFactory(new PostProcessingTypeAdapter())
            .create();
    }
}

File: app/src/test/kotlin/fr/free/nrw/commons/auth/login/UserExtendedInfoClientTest.kt
package fr.free.nrw.commons.auth.login

import android.net.Uri
import com.google.gson.GsonBuilder
import com.google.gson.stream.MalformedJsonException
import fr.free.nrw.commons.MockWebServerTest
import io.reactivex.observers.TestObserver
import org.junit.Before
import org.junit.Test
import fr.free.nrw.commons.wikidata.model.WikiSite
import fr.free.nrw.commons.wikidata.mwapi.MwQueryResponse
import fr.free.nrw.commons.wikidata.json.NamespaceTypeAdapter
import fr.free.nrw.commons.wikidata.json.PostProcessingTypeAdapter
import fr.free.nrw.commons.wikidata.json.UriTypeAdapter
import fr.free.nrw.commons.wikidata.json.WikiSiteTypeAdapter
import fr.free.nrw.commons.wikidata.model.page.Namespace
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory

class UserExtendedInfoClientTest : MockWebServerTest() {
    private var apiService: LoginInterface? = null
    private val observer = TestObserver<MwQueryResponse>()
    private val gson = GsonBuilder()
        .registerTypeHierarchyAdapter(Uri::class.java, UriTypeAdapter()
            .nullSafe())
        .registerTypeHierarchyAdapter(
            Namespace::class.java, NamespaceTypeAdapter()
            .nullSafe())
        .registerTypeAdapter(
            WikiSite::class.java, WikiSiteTypeAdapter()
            .nullSafe())
        .registerTypeAdapterFactory(PostProcessingTypeAdapter())
        .create()

    @Before
    @Throws(Throwable::class)
    override fun setUp() {
        super.setUp()

        apiService = Retrofit.Builder()
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .addConverterFactory(GsonConverterFactory.create(gson))
            .baseUrl(server().url)
            .build()
            .create(LoginInterface::class.java)
    }

    @Test
    @Throws(Throwable::class)
    fun testRequestSuccess() = runTest {
        enqueueFromFile("user_extended_info.json")

        val result = apiService!!.getUserInfo("USER")

        val userInfo = result.body()?.query()?.userInfo()
        assertEquals(24531888, userInfo?.id())
        val userResponse = result.body()?.query()?.getUserResponse("USER")?.name()
        assertEquals("USER", userResponse)
    }

    @Test
    fun testRequestResponse404() = runTest {
        enqueue404()

        val result = apiService!!.getUserInfo("USER")

        assertFalse(result.isSuccessful)
        assertEquals(404, result.code())
    }

    @Test
    @Throws(Throwable::class)
    fun testRequestResponseMalformed() = runTest {
        enqueueMalformed()

        try {
            apiService!!.getUserInfo("USER")
            fail("Exception expected")
        } catch (e: Exception) {
            assertTrue(e is MalformedJsonException)
        }
    }
}
